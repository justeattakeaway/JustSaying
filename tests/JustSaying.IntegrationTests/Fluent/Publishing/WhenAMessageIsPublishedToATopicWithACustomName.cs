using JustSaying.Messaging.MessageHandling;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Publishing;

public class WhenAMessageIsPublishedToATopicWithACustomName(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Message_Is_Handled()
    {
        // Arrange
        var handler = new InspectableHandler<SimpleMessage>();

        var services = GivenJustSaying()
            .ConfigureJustSaying((builder) =>
                builder
                    .Publications((options) =>
                        options.WithTopic<SimpleMessage>(configure => { configure.WithTopicName("my-special-topic"); }))
                    .Subscriptions((options) =>
                        options.ForTopic<SimpleMessage>("my-special-topic",
                            subscriptionBuilder => { subscriptionBuilder.WithQueueName(UniqueName); })))
            .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);

        string content = Guid.NewGuid().ToString();

        var message = new SimpleMessage()
        {
            Content = content
        };

        string json = "";

        await WhenAsync(
            services,
            async (publisher, listener, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                var listenerJson = JsonConvert.SerializeObject(listener.Interrogate(), Formatting.Indented);
                var publisherJson = JsonConvert.SerializeObject(publisher.Interrogate(), Formatting.Indented);

                await publisher.PublishAsync(message, cancellationToken);

                json = string.Join($"{Environment.NewLine}{Environment.NewLine}",
                        listenerJson,
                        publisherJson)
                    .Replace(UniqueName, "integrationTestQueueName", StringComparison.Ordinal);

                await Patiently.AssertThatAsync(OutputHelper,
                    () =>
                        handler.ReceivedMessages.Any(x => x.Content == content).ShouldBeTrue());
            });

        json.ShouldMatchApproved(opt => opt.SubFolder("Approvals"));
    }
}
