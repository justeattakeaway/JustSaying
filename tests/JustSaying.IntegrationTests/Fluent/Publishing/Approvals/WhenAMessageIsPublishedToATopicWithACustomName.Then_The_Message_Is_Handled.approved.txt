{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": [
      {
        "MessageType": "SimpleMessage",
        "QueueName": "integrationTestQueueName",
        "MiddlewareChain": [
          "MessageContextAccessorMiddleware",
          "LoggingMiddleware",
          "StopwatchMiddleware",
          "SqsPostProcessorMiddleware",
          "ErrorHandlerMiddleware",
          "HandlerInvocationMiddleware`1[JustSaying.TestingFramework.SimpleMessage]"
        ]
      }
    ]
  },
  "PublishedMessageTypes": {},
  "SubscriptionGroups": {
    "Groups": [
      {
        "Name": "integrationTestQueueName",
        "ConcurrencyLimit": 10,
        "Multiplexer": {
          "ChannelCapacity": 100,
          "ReaderCount": 1
        },
        "ReceiveBuffers": [
          {
            "BufferSize": 10,
            "QueueName": "integrationTestQueueName",
            "Region": "eu-west-1",
            "Prefetch": 10
          }
        ]
      }
    ]
  }
}

{
  "Region": "eu-west-1",
  "Middleware": {
    "Middlewares": []
  },
  "PublishedMessageTypes": {
    "SimpleMessage": {
      "Arn": "arn:aws:sns:us-east-1:000000000000:my-special-topic"
    }
  },
  "SubscriptionGroups": {
    "Groups": []
  }
}